v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/endtoend/EndToEndTestingITBase.java

```diff
@@ -0,0 +1,449 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(EndToEndTestingITBase.class);
+  private static FlexTemplateDataflowJobResourceManager flexTemplateDataflowJobResourceManager;
+  public DatastreamResourceManager datastreamResourceManager;
+  protected JDBCSource jdbcSource;
+
+  protected SpannerResourceManager createSpannerDatabase(String spannerSchemaFile)
+      throws IOException {
+    SpannerResourceManager spannerResourceManager =
+        SpannerResourceManager.builder("e2e-main-" + testName, PROJECT, REGION)
+            .maybeUseStaticInstance()
+            .build();
+
+    String ddl;
+    try (InputStream inputStream =
+        Thread.currentThread().getContextClassLoader().getResourceAsStream(spannerSchemaFile)) {
+      if (inputStream == null) {
+        throw new FileNotFoundException("Resource file not found: " + spannerSchemaFile);
+      }
+      try (BufferedReader reader =
+          new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))) {
+        ddl = reader.lines().collect(Collectors.joining("\n"));
+      }
+    }
+
+    if (ddl.isBlank()) {
+      throw new IllegalStateException("DDL file is empty: " + spannerSchemaFile);
+    }
+
+    String[] ddls = ddl.trim().split(";");
+    for (String d : ddls) {
+      d = d.trim();
+      if (!d.isEmpty()) {
+        spannerResourceManager.executeDdlStatement(d);
+      }
+    }
+    return spannerResourceManager;
+  }
+
+  protected SpannerResourceManager createSpannerMetadataDatabase() {
+    SpannerResourceManager spannerMetadataResourceManager =
+        SpannerResourceManager.builder("e2e-meta-" + testName, PROJECT, REGION)
+            .maybeUseStaticInstance()
+            .build();
+    String dummy = "create table t1(id INT64 ) primary key(id)";
+    spannerMetadataResourceManager.executeDdlStatement(dummy);
+    return spannerMetadataResourceManager;
+  }
+
+  public PubsubResourceManager setUpPubSubResourceManager() throws IOException {
+    return PubsubResourceManager.builder(testName, PROJECT, credentialsProvider).build();
+  }
+
+  public SubscriptionName createPubsubResources(
+      String identifierSuffix,
+      PubsubResourceManager pubsubResourceManager,
+      String gcsPrefix,
+      GcsResourceManager gcsResourceManager,
+      String mode) {
+    String topicNameSuffix = mode + "-it" + identifierSuffix;
+    String subscriptionNameSuffix = mode + "-it-sub" + identifierSuffix;
+    TopicName topic = pubsubResourceManager.createTopic(topicNameSuffix);
+    SubscriptionName subscription =
+        pubsubResourceManager.createSubscription(topic, subscriptionNameSuffix);
+    String prefix = gcsPrefix;
+    if (prefix.startsWith("/")) {
+      prefix = prefix.substring(1);
+    }
+    if (mode == "rr") {
+      prefix += "/retry/";
+    }
+    gcsResourceManager.createNotification(topic.toString(), prefix);
+    return subscription;
+  }
+
+  protected void createAndUploadShardConfigToGcs(
```

Comment:
Is this shard config specific to Reverse replication? If so we should include ReverseReplication in the name of the method. like "createAndUploadReverseReplicationShardConfigToGcs"

Reply:
It is. Sure let me doi that